src/device/net/dev_raw.rs

```diff
@@ -0,0 +1,302 @@
+use super::{Config, EthernetAddress, Features, VirtioNetHdr};
+use super::{MIN_BUFFER_LEN, NET_HDR_SIZE, QUEUE_RECEIVE, QUEUE_TRANSMIT};
+use crate::hal::Hal;
+use crate::queue::VirtQueue;
+use crate::transport::Transport;
+use crate::volatile::volread;
+use crate::{Error, Result};
+use log::{debug, info, warn};
+use zerocopy::AsBytes;
+
+/// Raw driver for a VirtIO block device.
+///
+/// This is a raw version of the VirtIONet driver. It provides non-blocking
+/// methods for transmitting and receiving raw slices, without the buffer
+/// management. For more higher-level fucntions such as receive buffer backing,
+/// see [`VirtIONet`].
+///
+/// [`VirtIONet`]: super::VirtIONet
+pub struct VirtIONetRaw<H: Hal, T: Transport, const QUEUE_SIZE: usize> {
+    transport: T,
+    mac: EthernetAddress,
+    recv_queue: VirtQueue<H, QUEUE_SIZE>,
+    send_queue: VirtQueue<H, QUEUE_SIZE>,
+}
+
+impl<H: Hal, T: Transport, const QUEUE_SIZE: usize> VirtIONetRaw<H, T, QUEUE_SIZE> {
+    /// Create a new VirtIO-Net driver.
+    pub fn new(mut transport: T) -> Result<Self> {
+        let negotiated_features = transport.begin_init(Features::MAC | Features::STATUS);
```

Comment:
Why not `RING_EVENT_IDX`, which was previously included in `SUPPORTED_FEATURES`?

Reply:
Previous code written by @equation314 did not contain the RING_EVENT_IDX feature, after a conversation with him, it will be added to SUPPORTED_FEATURES soon.